cpp/src/groupby/hash_groupby.cuh

```diff
@@ -80,6 +82,23 @@ struct is_same_functor : std::false_type{};
 template <template <typename> class T>
 struct is_same_functor<T,T> : std::true_type{};
 
+/**---------------------------------------------------------------------------*
+ * @brief Functor for `typed_groupby` to use with `type_dispatcher`
+ * 
+ *---------------------------------------------------------------------------**/
+template <template <typename T> class op>
+struct typed_groupby_functor
+{
+  template <typename TypeAgg, typename... Ts>
+  gdf_error operator()(Ts&&... args)
+  {
+    // TypeAgg can a wrapped type (e.g detail::wrapper<gdf_date32, GDF_DATE32>)
+    // in which case, we want to pass the underlying type (gdf_date32 = int)
+    using TypeAggUndl = cudf::detail::unwrapped_type_t<TypeAgg>;
```

Comment:
Why is this necessary? 

Ideally you shouldn't need to access the underyling type of the wrapper. That's considered to be "hidden" information.

Reply:
I tried but it broke at a call to `atomicCAS`.

Reply:
In that case, what we'll need to do is to add some overloads for `atomicCAS` for the wrapper types.

Inside the `atomicCAS` overloads, you can access the underlying `value` and pass it through the "normal" `atomicCAS`.

I believe @kovaltan is doing something similar in https://github.com/rapidsai/cudf/pull/1005 and you could coordinate that effort with him.

Reply:
Sorry, I also use underlying type of the wrapper #1005 
I think using  underlying type  for cuda kernels is much simpler.

https://github.com/rapidsai/cudf/blob/e68dc4353db5c774b0c3a9bc651a4f4822910306/cpp/src/reductions/reductions.cu#L112-L119

Reply:
We really don't want to do that, the underlying type should not be used directly here. It should only be used inside `wrapper` operators and specializations of `atomic*` operations.

The reason for this is because some of the wrapper types may have different semantics than their underlying type. Perfect example is `cudf::bool8` uses a `int8_t` as the underlying type. However `cudf::bool8{1} + cudf::bool8{1} == 1`, if you were to use the underlying type directly, you would produce the incorrect results of `1 + 1 = 2`.

You should think of the `wrapper` types as opaque objects that you should not be accessing the underlying type nor value.

The only entities that should know about the underlying type or value are `wrapper` operators and other operators that are specialized to work on the `wrapper` types, e.g., `atomic*` operations.

Reply:
To be clear, I'm highly averse to this trait. It encourages treating the `wrapper` types as something other than opaque types. 

This is the same reason that the use of `unwrap` is highly discouraged.

Reply:
Thanks to @kovaltan's atomics, I was able to refactor `hash_groupby` without using the trait. But to fix that, I had to make a clone of `std::numeric_limits` that works on wrapper types. That, however, needs that trait. I agree that this trait should not be used by functionality code and that's why it's inside the `detail` namespace. Ideally I intend this to be useful to utility writers.